src/world.js

```diff
@@ -136,7 +141,14 @@ class World extends EventEmitter {
     // interval. The set structure is maintaining the order of insertion
     for (const [key, { chunkX, chunkZ }] of this.savingQueue.entries()) {
       this.finishedSaving = Promise.all([this.finishedSaving,
-        this.storageProvider.save(chunkX, chunkZ, this.columns[key])])
+        this.storageProvider.save(chunkX, chunkZ, this.columns[key])
+          .then(() => {
+            if (this.unloadQueue.has(key)) {
+              this.unloadQueue.delete(key)
+              this.unloadColumn(chunkX, chunkZ)
```

Comment:
this works thanks to the bug above
can you instead make a function for
      delete this.columns[key]
      const columnCorner = new Vec3(chunkX * 16, 0, chunkZ * 16)
      this.emit('chunkColumnUnload', columnCorner)
and call it here ?

Reply:
Ok so let me explain more here.
You are calling unloadChunk methods
There there is 2 branch
First branch is adding to unload queue, that happens only if the saving queue contains that key
Second is actually unloading the column

Here we are going in the second branch because the saving queue has already been emptied.
However we are still in the middle of the loop so this shouldn't have happened yet, the saving queue should not be empty

The reason why the saving queue is empty in that then is that the callback here is executed *after the saving queue emptying* where my second comment is placed.

This is not right, we should first save all the items and only then empty the saving queue.

That's why I'm suggesting to put an await at the line below to wait all the saving have been done before emptying the queue

Reply:
There is however then a last problem to solve : the saveNow function should be executed only once, it doesn't make sense to have 2 of them running concurrently
This can be protected either inside the function with a promise to be awaited or outside, at the place of the set interval

Reply:
That's what I meant by "Is there a risk in the current savingQueue implementation that if the savingQueue is not processed within the startSaving interval, the same chunk will be saved multiple times?"

It seems to me that the easiest fix for that would be to change setInterval to setTimeout, and have it call itself.

Reply:
What about using .delete() after a savingQueue entry has been handled, rather than .clear()? Couldn't .clear() end up removing mapping entries that were added after saveNow started but before it finished?